File: example/js/automate.js
class AutomateString{
	chars = []
	get length(){
		return this.chars.length;
	}
	constructor(str){
		this.chars = str.split('')
	}
	cursor = 0
	next(){
		return this.cursor>=this.length?null:this.chars[this.cursor++]
	}
	last(){
		return this.chars[this.cursor-1]
	}
	hasNext(){
		return this.cursor<=this.length&&(this.chars[this.cursor])
	}
	forEach(callback){
		var c = this.next()
		while(this.hasNext()){
			callback(c,this)
			c = this.next()
		}
		callback(c,this)
		callback(null,this)
		return 
	}
	static concat(...args){var e = new AutomateString("");e.chars = args;return e}
}
Array.prototype.last = function(){
	return this[this.length-1]
}
class AutomateBase {
	static ETAT_INITIAL = 0
	static REJECT = false
	static ACCPET = true
	translate(entree){
		if(!(entree instanceof AutomateString))
			entree = new AutomateString(entree)
		return entree
	}
}
class AutomateBuilder extends AutomateBase{
	END = -3
	ALL = -2
	ME = -1
	#reconizeToken = []
	addToken(input,output,stats = [0],after = [],){
		var token = {input,output,stats}
		token.stat = this.#reconizeToken.length
		if(stats.includes(-1)){
			token.stats.push(token.stat)
		}
		this.#reconizeToken.push(token)
		after.forEach(n=>{
			if(n>0 && n < this.#reconizeToken.length)
				this.#reconizeToken[n].stats.push(token.stat)
		})
		return token.stat
	}
	getToken(token,stat){
		var find = this.#reconizeToken.find(t=>{
			var i = t.input
			// if(t.output == "htmlcontent"){
			// 	console.log(t.stats,stat)
			// }
			if( ! t.stats.includes(stat) && !t.stats.includes(-2))
				return false
			if(i instanceof RegExp){
				return i.test(token)&&! token.split(i).join('').length
			}else{
				return i == token
			}
		})
		// if(find)
			// console.log(find,find.stats.includes(stat))
		return find
	}
	recognize(entree){
		entree = this.translate(entree)
		var stat = AutomateBase.ETAT_INITIAL
		var tokensLS = []
		var culsym = ""
		var token;
		// var c = entree.next()
		entree.forEach(c=>{
		if(token = this.getToken(culsym,stat)){ 
				// console.log(token,stat)
				tokensLS.push({
					input:culsym,
					output:token.output,
					stat:token.stat,
					token,
				})
				stat = token.stat
				culsym = ""
			}
			if(c)
				culsym += c
		})
		tokensLS.push({
			input:culsym,
			output:-1,
			stat,
		})
		tokensLS.clean = function(){
			return this.reduce((p,o)=>{
				if(o.input.length>0)p.push(o.input)
				return p
			},[])
		}
		return tokensLS
	}
	regroup(text,groups = []){
		var parsed = this.recognize(text)
		groups = groups.map(x=>x.map(y=>this.#reconizeToken[y]||y))
		var output = ""
		var t = ""
		var s = 0
		var n = 1
		var group = []
		var regroup = []
		text = parsed.reduce((p,o)=>{
			var g = groups.find((g)=>g.includes(output))
			var gn = groups.find((g)=>g.includes(o.output))
// gn == undefined ? output == o.output : g==gn
			if( gn != undefined && g==gn ){
				t += o.input
				s = s+o.stat
				regroup.push(o)
				n++
			}else{
				p.push({
					input:t,
					stat:s/n,
					n,
					regroup
				})
				t = o.input
				s = o.stat
				n = 1
				regroup = [o]
			}
			output = o.output

			return p
		},[])
		text.push({
			input:t,
			stat:s/n,
			n
		})
		text.clean = function(){
			return this.reduce((p,o)=>{
				if(o.input.length>0)p.push(o.input)
				return p
			},[])
		}
		return text
	}
	print(text,group){
		var parsed = this.regroup(text,group)
		var text = ""
		parsed.forEach(token=>{
			text += token.input.padStart(token.stat,"\t")+"\n"
		})
		return text
	}
	printExecution(text,group){
		var parsed = this.regroup(text,group)
		var text = ""
		var ident = 0
		parsed.forEach((token,n)=>{
			text += "".padStart(ident,"\t")+token.input+"\n"
			console.log(token)
			if(token.regroup && token.regroup.length){
				if(
					token.regroup[0].token.stats.includes(0) &&
					token.regroup[0].token.stats.length == 0
					){
					ident = 0
				} else if (
					token.regroup[0].token.stats.length == 1
					){
				} else if (
					! token.regroup[0].token.stats.includes(-1)
					){
					if(token.regroup[0].token.stats.includes(0))
						ident ++
				}else{
					ident++
				}
			}else{
				console.log(parsed[n+1] , parsed[n+1].regroup)
				if(parsed[n+1] && parsed[n+1].regroup && parsed[n+1].regroup.last().token.stats.includes(AutomateBuilder.END))
					ident--
			}
		})
		return text
	}
	execute(text,groups){
		var parsed;
		if(groups){
			parsed = this.regroup(text,groups).clean()
		}else{
			parsed = this.recognize(text).clean()
		}
		console.log(parsed)
	}
}

File: src/main/java/automate/Automate.java
package automate;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.regex.Pattern;
import java.util.stream.Stream;

public class Automate extends AutomateBase{
	final public static int ALL = -2;
	final public static int ME = -1;
	private List<Token> reconizeToken = new ArrayList<Token>();
	private Token getToken(String token,int stat){
		Stream<Token> stream = this.reconizeToken.stream();
		//recherche des tokens
		stream= stream.filter((Token t)->{
			//recheche des tokens qui peuvent existé après cette etat
			var c = !(t.includes(stat) && !t.includes(ALL));
			if( c )
				return false;
			//  et si le text peut correspondre au token
			return t.test(token);
		});
		Optional<Token> optionalToken = stream.findFirst();
		if( ! optionalToken.isEmpty())
			return optionalToken.get();// retoune le token si il existe
		return null;
	}
	private boolean debug = false;
	public void setDebug(boolean mode){
		debug = mode;
	}
	private Token token;
	private int stat;
	private String culsym;
	/**
	 * Analyse Lexical : 
	 * cette methode reconnait les tokens dans un text
	 * @param entree
	 * @return
	 */
	public List<RegonizeToken> recognize(AutomateString entree){
		stat = AutomateBase.ETAT_INITIAL;
		culsym = "";
		List<RegonizeToken> tokensLS = new ArrayList<RegonizeToken>();
		// pour chaque symbole de la chaine
		entree.forEach((Character c)->{
			//cherche le token dans les token connu
		token = this.getToken(culsym,stat);
		if(debug){
			System.out.println("input : '"+c+"' or "+((int) c));
			if(token == null){
				System.out.println("no token find at stat : "+stat+"; next token are : "+culsym);
			}else{
				System.out.println("token find : "+token+" stat : "+token.getStat());
			}
		}
		// si le token est connu
		if(token != null){ 
				// token reconnu = (entre utilisateur,nom etat,etat,token original)
				RegonizeToken regonize = new RegonizeToken(
					culsym,token.getOutput(),token.getStat(),token
					);
				tokensLS.add(regonize);
				stat = token.getStat();//on passe a l'etat suivant
				culsym = "";
			}
			if(c != null)
				culsym += c;//quand le text du token n'est pas connu on y ajoute le prochain symbole
		},(Character c)->{
			// a la fin en registre le dernier token non reconnu
			tokensLS.add(new RegonizeToken(culsym,stat));
		});
		return tokensLS;
	}
	/**
	 * Analyse lexical (avec regroupement)
	 * @see recognize
	 * @param text
	 * @param groups
	 * regroupe les tokens selon les groupes données
	 * @return liste de tokens
	 */
	public List<RegonizeToken> regroup(AutomateString text,List<List<Token>> groups){
		List<RegonizeToken> parsed = this.recognize(text);
		List<RegonizeToken> tokenLs = new ArrayList<RegonizeToken>();
		token = null;
		String t = "";
		String _out="";
		int s = 0;
		RegonizeToken last = null;
		List<RegonizeToken> regroup = new ArrayList<>();
		for (RegonizeToken regonizeToken : parsed) {// pour chaqque token reconnu
			List<Token> g = null;
			List<Token> gn = null;
			Optional<List<Token>> opt;
			// trouver le premier group conserner avec l'ancient token
			opt = groups.stream().filter((List<Token> gls)->{return gls.contains(token);}).findFirst();
			if(! opt.isEmpty()){
				g = opt.get();// g = group avec le token precende
			}
			// trouver le premier group conserner avec le nouveau token
			opt = groups.stream().filter((List<Token> gls)->gls.contains(regonizeToken.getToken())).findFirst();
			if(! opt.isEmpty()){
				gn = opt.get();//gn = group avec le nouveau token
			}
			if(gn != null && g==gn){//si le token a un group et que c'est le meme que le group precedent
				t += regonizeToken.getInput();// on group les entres
				regroup.add(regonizeToken);//on regroup les token
			}else{
				// Token token_correction = this.getBestToken(t,last.getToken());
				// if(token_correction != null){
				// 	s = token_correction.getStat();
				// 	_out = token_correction.getOutput();
				// }
				// if(regroup != null){

				// }
				last = new RegonizeToken(t, _out, s,token);
				tokenLs.add(last);
				t = regonizeToken.getInput();
				_out = regonizeToken.getOutput();
				s = regonizeToken.getStat();// changement etat
				regroup.clear();// creer un nouveau group
				regroup.add(regonizeToken);
			}
			token = regonizeToken.getToken();
		}
		tokenLs.add(new RegonizeToken(t,-1));
		return tokenLs;
	}
	/**
	 * cette methode était sensé traduire un text donnée
	 * regroupé les parties donnée
	 * puis en faire l'execution
	 */
	/*
	execute(String text,int[] regroup,actions){
		var groups = this.regroup(text,regroup)
		// console.log(groups)
		groups.forEach(x=>{
			var fx = x.output
			fx = typeof fx == "function"?fx:actions[fx]
			fx = fx||actions.default
			if(fx && x){
				fx(x.input,x)
			}else{
				console.log(x.input)
			}
		})
	}
	*/
	/** == Methode de debug == */
	/**
	 * BestToken (DEBUG)
	 * trouver le token qui correspond au text entre en parametre qui peut suivre le token last
	 * @param token
	 * @param last
	 * @return
	 */
	public Token getBestToken(String token,Token last){
		//les tokens qui correspond au text entre en parametre
		Stream<Token> proba = this.showToken(token).filter((Token t)->{
			return t.includes(last.getStat());//peut suivre le  token last
		});
		//les tokens qui correspond au text entre en parametre
		proba = proba.filter(t->{
			Object i = t.getInput();
			if(i instanceof Pattern ){
				return ((Pattern)i).matcher(token).start() == 0;
			}else if(i instanceof String){
				return ((String) i) == token;
			}else{
				return false;
			}
		});
		Optional<Token> optionalToken = proba.findFirst();
		if( ! optionalToken.isEmpty())
			return optionalToken.get();
		return null;
	}
	/**
	 * showToken (DEBUG)
	 * trouver les tokens avec un etat (entier)
	 * @param token
	 * @return
	 */
	public Stream<Token> showToken(int token){
		return this.reconizeToken.stream().filter(t->{
			return t.getStat() == token;
		});
	}
	/**
	 * showToken (DEBUG)
	 * trouver les tokens qui correspond au text entre en parametre
	 * @param token
	 * @return
	 */
	public Stream<Token> showToken(String token){
		return this.reconizeToken.stream().filter(t->{
			var i = t.getInput();
			if(i instanceof Pattern ){
				return ((Pattern)i).matcher(token).start() == 0;
			}else if(i instanceof String){
				return ((String) i) == token;
			}else{
				return false;
			}
		});
	}
	/**
	 * Ajouter un token a la liste des token reconnu par l'automate
	 * il defini aussi un etat pour se token
	 * @param token
	 */
	public void addToken(Token token){
		token.setStat(this.reconizeToken.size()+1);
		this.reconizeToken.add(token);
	}
	
}
